--- qemu-kvm-1.2.0/configure~	2013-01-30 14:21:52.705707010 +0100
+++ qemu-kvm-1.2.0/configure	2013-01-30 15:51:18.582483375 +0100
@@ -2678,13 +2678,14 @@
 cat > $TMPC <<EOF
 #include <signal.h>
 #include <time.h>
-int main(void) { return clock_gettime(CLOCK_REALTIME, NULL); }
+int main(void) { timer_t tid; (void)timer_gettime(tid, NULL); return clock_gettime(CLOCK_REALTIME, NULL); }
 EOF
 
 if compile_prog "" "" ; then
   :
 elif compile_prog "" "-lrt" ; then
   LIBS="-lrt $LIBS"
+  libs_qga="-lrt $libs_qga"
 fi
 
 if test "$darwin" != "yes" -a "$mingw32" != "yes" -a "$solaris" != yes -a \
